File: design_word_search.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_word = False


class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root

        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.end_of_word = True

    def search(self, word: str) -> bool:
        def dfs(j, root):
            node = root

            for i in range(j, len(word)):
                c = word[i]

                if c == ".":
                    for child in node.children.values():
                        if dfs(i + 1, child):
                            return True
                    return False
                else:
                    if c not in node.children:
                        return False
                    node = node.children[c]
            return node.end_of_word

        return dfs(0, self.root)

File: designWordSearch.js
class TrieNode {
  constructor() {
    this.children = {};
    this.endOfWord = false;
  }
}

class WordDictionary {
  constructor() {
    this.root = new TrieNode();
  }

  /**
   * @param {string} word
   * @return {void}
   */
  addWord(word) {
    let node = this.root;

    for (let c of word) {
      if (!(c in node.children)) {
        node.children[c] = new TrieNode();
      }
      node = node.children[c];
    }
    node.endOfWord = true;
  }

  /**
   * @param {string} word
   * @return {boolean}
   */
  search(word) {
    const dfs = (j, root) => {
      let node = root;

      for (let i = j; i < word.length; i++) {
        let c = word[i];

        if (c === ".") {
          for (let child of Object.values(node.children)) {
            if (dfs(i + 1, child)) {
              return true;
            }
          }
          return false;
        } else {
          if (!(c in node.children)) {
            return false;
          }
          node = node.children[c];
        }
      }
      return node.endOfWord;
    };

    return dfs(0, this.root);
  }
}
